README
Overview
~~~~~~~~

This is gtk-g-ray2, a tool to access the Wintec WBT-201 from free
operating systems (but you can use it on proprietary ones too if you
like).

gtk-rays-2
~~~~~~~~~~

There is a "C" version (current) and an older Ruby version, which is
useful for experimentation.

To build the 'C' version:

$ ./configure [--prefix=/usr]
$ make
$ sudo make install

By default, device names of /dev/ttyUSB0 and bluetooth are
offered. You can change the devices available by editing the file
$HOME/.config/g-rays2/g-rays2rc

(gtk-rays-2 creates the file on first run if it does not exist).

e.g.

$ cat /home/jrh/.config/g-rays2/g-rays2rc
[g-rays2]
devices = /dev/ttyUSB0;/dev/rfcomm0

After the key "devices = " there is a semi-colon (;) separated list of
possible device names.

e.g.
devices = /dev/ttyUSB0;/dev/ttyUSB1;/dev/ttyUSB2;/dev/rfcomm0;/dev/rfcomm1

Any device name of just "bluetooth" or containing the string "rfcomm"
will be cause bluetooth discovery to find the real device (new in 2.00).

From version 0.99, the preferences dialogue allows the user to set and
modify the serial device list.

In order to save log files using gpsbabel, it is necessary (as of
2007-05-19) to:

* Use the CVS version;

If the [CVS] version is earlier than 2007-07-03, then you need to:

* Apply a patch to get it to use standard input as the serial device.

The patch file is included in this archive as docs/gpsbabel-fd0_as_serial.patch

To apply:

$ cd gpsbabel # Source directory
$ # gtk-g-rays2 unpacked in /tmp/
$ patch < /tmp/gtk-g-rays2-0.4/wbt-201/docs/gpsbabel-fd0_as_serial.patch
$ make
$ # Now install on path

Requirements
~~~~~~~~~~~~

gtk+ 2.8.0 or later
libglade
cairo

If cairo is not available, then the application will build, without
the graphical component.

MacOSX
~~~~~~

From Matthias Wiedemann:

 updating gtk, because my fink gtk2 version was 2.6.10
	macport could give me version 2.10.11
	I think you need at least gtk 2.8

 - also, I fetched cairo libraries

 - yeah, it works on Mac OS

ruby 
~~~~

The ruby version is no longer included in the distributed
tar file. Just ask if you want it.

Historically:

For the ruby version (in the ruby directory).

Usage: wbt201.rb [options]
    -d, --dump-gps                   false
    -v, --verbose                    false
    -h, --help                       Show this message

Please be advised that wbt201.glade must be in the working directory
for the ruby version. This is feature.

See www/index.html or http://www.daria.co.uk/gps/ for more
details.

(c) 2007-2013 Jonathan Hudson <jh+gps@daria.co.uk>


ChangeLog
2.05 2014-01-21
	Gtk+-3.0 clean up. New bluetooth solution ...

2.02 2013-02-03
	Reworked bluetooth code to improve performance and stability.
	Reinstated manual BT workarounds (for Robert Peter). Updated the
	manual and added some notes on debugging BT issues.

2.01 2011-12-04
	Updated the Bluetooth code (C rather than vala generated).

2.0  2011-11-27
	GTK+-3.0 version. Added auto mode for device detection.

1.23 2011-03-05
	"March Madness". Runs when built without deprecated set
	(-DGDK_DISABLE_DEPRECATED=1 -DG_DISABLE_DEPRECATED)

1.22 2011-01-18
	"Cold Snap". Builds without deprecated set
	(-DGDK_DISABLE_DEPRECATED=1 -DG_DISABLE_DEPRECATED)

1.21 2010-07-02
	"Home coming". Added use of /dev/gpsN option. Not perfect but good
	enough for my purposes.

	Added a Gentoo patch (Samuli Suominen) to allow compilation with
	CFLAGS="-O0"

1.20 2009-06-25
	"Summer daze". Added a default save directory option.

1.11 2008-06-18
	"To the point". Incorporated a patch by Torsten Hilbrich to set
	LC_NUMERIC to 'C' while parsing GPS data. This gives correct
	values for decimal fields in locales where the decimal separator
	is not '.'. Output remains locale specific.

1.10 2008-04-12
       "First Cuckoo of spring". Fixed bug in glade file. No code changes.

1.06 2007-11-12
       "Self starter". Added the ability to call a script to set up and
	tear down the serial bindings, primarily for Bluetooth. Please see
	the manual for a description of how this works. A sample script
	for /dev/rfcommX (g-rays2-rfcomm.sh) is provided.

1.05 2007-11-11
	"Babel". Added gettext framework Changed mktime() to timegm() for
	NMEA data (reported by Tilman Vogel)

	Added the Nokia N800 (maemo) as a build target, fully integrated
	into the source tree. This was made possible by the efforts of
	Tilman Vogel, who did the Maemo port for 1.04. The Maemo port
	requires Tilman's port of gpsbabel.

	Debian builds are available for x86_64 (amd64), ia32 (i386),
	powerpc and maemo (N800) platforms.

1.04 2007-09-11
	"Build freely". Builds in FreeBSD.

1.03 2007-09-10
	"Deer Park". Set the saved file name to default to the date.

1.02 2007-07-28
	"Après nous le déluge". (sort of missed changes for  1.0,1.01).
	Documentation.
	Control-C in the Data window copies position, altitude, timestamp
	and H/V/PDOP to clipboard.
	Builds on MacOSX, thanks to some assistance from Matthias
	Wiedemann.

0.99 2007-07-07
	"Twice Daily". Added a preferences dialogue that allows you to
	save the device list names, and indicate if you want to save the
	last used device for subsequent recall.
	Added options for saving GPX; tracks, waypoints or both.
	Removed the 'popt' dependency.

0.5 2007-07-07
	"Seven times seven". Updated the README after Axel Müller pointed
	out that I hadn't documented how to change the device list
	entries. Please also note that CVS gpsbabel has the patch to allow
	stdin as the serial device, so you no longer need to patch it.  **
	No code changes. **

0.4 2007-06-01
	"Back from Tokyo". Device are now read from a configuration file,
	~/.config/g-rays2/g-rays2rc (which is created if it doesn't
	exist).  Mainly implemented on VS901 Narita to London.

0.1 2007-05-19
	"Fit for public consumption". I hope so. First real release.



src/wbt201.c
/* -*- mode: C; coding: utf-8; -*-
 */

/*
 * Copyright (C) 2007-2013 Jonathan Hudson <jh+gps@daria.co.uk>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

#include <gtk/gtk.h>
#include <glib/gstdio.h>
#include <assert.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sys/types.h>
#include <unistd.h>
#include <libintl.h>
#include <fcntl.h>
#include <errno.h>
#include <gdk/gdkkeysyms.h>
#include <math.h>
#include <locale.h>

#ifdef __linux
#define DEFDEVS "[g-rays2]\ndevices = auto;bluetooth;/dev/ttyUSB0\n"
#elif defined(__FreeBSD__)
#define DEFDEVS "[g-rays2]\ndevices =/dev/cuaU0\n"
#else
#define DEFDEVS "[g-rays2]\ndevices = (undefined)\n"
#endif

#include "wbt201.h"

#ifndef PACKAGE_DATA_DIR
# define PACKAGE_DATA_DIR "."
#endif

#ifndef PACKAGE_LOCALE_DIR
# define PACKAGE_LOCALE_DIR "."
#endif

wbt201_t *wbt;

void message_box(const char *op, const char *s)
{
    GtkWidget *dialog;
    dialog = gtk_message_dialog_new (NULL,
                                     GTK_DIALOG_DESTROY_WITH_PARENT,
                                     GTK_MESSAGE_ERROR,
                                     GTK_BUTTONS_CLOSE,
                                     _("Error %s : %s"),
                                     op, s);
    gtk_dialog_run (GTK_DIALOG (dialog));
    gtk_widget_destroy (dialog);
}

void saveprefs(void)
{
    int fd;
    char **d;
    GString *g;
    g = g_string_new("[g-rays2]\ndevices=");


    g_mkdir_with_parents(wbt->confdir,0700);
    fd = g_open(wbt->confnam, O_WRONLY|O_CREAT|O_TRUNC, 0600);
    if(fd != -1)
    {
        for(d = wbt->devs; (d && *d); d++)
        {
            if(**d && !g_ascii_isspace (**d))
            {
                g_string_append(g, *d);
                g_string_append_c(g,';');
            }
        }
        *(g->str+g->len-1) = '\n';

        if(wbt->lastdev)
        {
            g_string_append(g, "lastdev=");
            g_string_append(g, wbt->lastdev);
            g_string_append_c(g,'\n');
        }

        if(wbt->usedevgps)
        {
            g_string_append(g, "usedevgps=true\n");
        }

        if(wbt->savepath)
        {
            g_string_append(g, "savepath=");
            g_string_append(g, wbt->savepath);
            g_string_append_c(g,'\n');
        }

        if(wbt->trackmarks)
        {
            g_string_append(g, "trackmarks=true\n");
        }

        if(wbt->defbtaddr)
        {
            g_string_append(g, "btaddr=");
            g_string_append(g, wbt->defbtaddr);
            g_string_append_c(g,'\n');
        }
        else if(wbt->btaddr)
        {
            g_string_append(g, "btaddr=");
            g_string_append(g, wbt->btaddr);
            g_string_append_c(g,'\n');
        }

        if (write(fd,g->str,g->len))
            ;  /* fail, silently to shutup older gcc */
        close(fd);
        g_string_free(g,TRUE);
    }
    else
    {
        const char *s= g_strerror (errno);
        message_box(_("writing preferences"),s);
    }
}

static void getprefs(void)
{
    gboolean res = FALSE;
    GKeyFile *kf = g_key_file_new();
    gsize len;

    if(wbt->confdir == NULL)
        wbt->confdir = g_build_filename(g_get_user_config_dir(), "g-rays2", NULL);
    if(wbt->confnam == NULL)
        wbt->confnam = g_build_filename(wbt->confdir, "g-rays2rc",NULL);

    if(wbt->confnam)
    {
        res = g_key_file_load_from_file(kf, wbt->confnam, G_KEY_FILE_NONE, NULL);
    }

    if (res == FALSE)
    {
        g_key_file_load_from_data(kf, DEFDEVS, sizeof(DEFDEVS)-1,
                                  G_KEY_FILE_NONE, NULL);
    }

    wbt->devs = g_key_file_get_string_list(kf,"g-rays2","devices",
                                                 &len, NULL);
    wbt->lastdev =  g_key_file_get_string(kf,"g-rays2","lastdev", NULL);
    wbt->usedevgps =  g_key_file_get_boolean(kf,"g-rays2","usedevgps", NULL);
    wbt->savepath = g_key_file_get_string(kf,"g-rays2","savepath", NULL);
    wbt->trackmarks = g_key_file_get_boolean(kf,"g-rays2","trackmarks", NULL);
    wbt->defbtaddr = g_key_file_get_string(kf,"g-rays2","btaddr", NULL);
    wbt->defrfcom  = g_key_file_get_string(kf,"g-rays2","rfcom", NULL);

    if(res == FALSE)
    {
        saveprefs();
    }
}

char *find_gps_dev(void)
{
    int n;
    static char gdev[]="/dev/gpsX";
    char *rgdev=NULL;

    for(n = 0; n < 10; n++)
    {
        gdev[8] = '0' + n;
        if(access(gdev, R_OK|W_OK) == 0)
        {
            rgdev = gdev;
            break;
        }
    }
    return rgdev;
}


void makecombo(void)
{
    GtkWidget *w;
    char **d;
    int lastn = 0;
    int n;
    if(wbt->devcombo == NULL)
    {
        assert(NULL != (w = GTK_WIDGET (gtk_builder_get_object (wbt->builder, "table1"))));
        wbt->devcombo = gtk_combo_box_text_new();
        gtk_grid_attach(GTK_GRID(w),wbt->devcombo,1,0,1,1);
        gtk_widget_show(wbt->devcombo);
    }
    else
    {
        for(n = wbt->ndevs-1; n >= 0; n--)
        {
            gtk_combo_box_text_remove (GTK_COMBO_BOX_TEXT(wbt->devcombo),n);
        }
    }

    for(n = 0, d = wbt->devs; (d && *d); d++)
    {
        if(**d && !g_ascii_isspace (**d))
        {
            gtk_combo_box_text_append_text(GTK_COMBO_BOX_TEXT(wbt->devcombo),*d);
            if(wbt->lastdev && 0 == strcmp(*d, wbt->lastdev))
            {
                lastn = n;
            }
            n++;
        }
    }


    if (wbt->usedevgps)
    {
        int ndev=-1;
        int dn;

        char *gdev = find_gps_dev();
        if (gdev)
        {
            for(dn = 0, d = wbt->devs; (d && *d); d++)
            {
                if(**d && !g_ascii_isspace (**d))
                {
                    if(0 == strcmp(*d, gdev))
                    {
                        ndev = dn;
                        break;
                    }
                    dn++;
                }
            }
            if(ndev != -1)
            {
                lastn = ndev;
            }
            else
            {
                gtk_combo_box_text_append_text(GTK_COMBO_BOX_TEXT(wbt->devcombo),gdev);
                lastn = n;
                n++;
            }
        }
    }
    wbt->ndevs = n;
    gtk_combo_box_set_active(GTK_COMBO_BOX(wbt->devcombo), lastn);
}

gboolean Accel(GtkAccelGroup *accel_group, GObject *acceleratable,
               guint keyval, GdkModifierType modifier)
{
    if (wbt->serfd > 0)
    {
        GtkWidget *w;
        assert(NULL != (w=GTK_WIDGET (gtk_builder_get_object (wbt->builder,
                                                              "notebook1"))));
        if(gtk_notebook_get_current_page (GTK_NOTEBOOK(w)) == 1)
        {
            const gchar *p;

            assert(NULL != (w=GTK_WIDGET (gtk_builder_get_object (wbt->builder,
                                                                  "utc"))));
            p = gtk_entry_get_text(GTK_ENTRY(w));
            if(p && strlen(p))
            {
                char buf[512];
                char *s;
                s = stpcpy(buf, p);
                *s++ = '|';
                assert(NULL != (w=GTK_WIDGET (gtk_builder_get_object (
                                                  wbt->builder,"lat"))));
                p = gtk_entry_get_text(GTK_ENTRY(w));
                s = stpcpy(s, p);
                *s++ = ' ';
                assert(NULL != (w=GTK_WIDGET (gtk_builder_get_object (
                                                  wbt->builder,"lon"))));
                p = gtk_entry_get_text(GTK_ENTRY(w));
                s = stpcpy(s, p);
                *s++ = '|';
                assert(NULL != (w=GTK_WIDGET (gtk_builder_get_object (
                                                  wbt->builder,"alt"))));
                p = gtk_entry_get_text(GTK_ENTRY(w));
                s = stpcpy(s, p);
                double rms;
                rms = sqrt(((wbt->gps.hdop*wbt->gps.hdop) +
                            (wbt->gps.vdop*wbt->gps.vdop) +
                            (wbt->gps.pdop*wbt->gps.pdop))/3.0);
                sprintf(s,"|%.1f\n", rms);
                gtk_clipboard_set_text(
                    gtk_clipboard_get(GDK_SELECTION_PRIMARY),
                    buf, strlen(buf));
                gtk_clipboard_store(gtk_clipboard_get(GDK_SELECTION_PRIMARY));
            }
        }
    }
    return TRUE;
}


static void make_accel(void)
{
    GtkWidget *w;
    GtkAccelGroup *accel = gtk_accel_group_new();
    GClosure *closure;
    closure = g_cclosure_new(G_CALLBACK(Accel),NULL,NULL);
    gtk_accel_group_connect(accel,GDK_KEY_C,GDK_CONTROL_MASK, 0, closure);
    assert(NULL != (w = GTK_WIDGET (gtk_builder_get_object (
                                        wbt->builder, "window1"))));
    gtk_window_add_accel_group(GTK_WINDOW(w),accel);
}

static void setup(void)
{
    GtkWidget *w;
    GtkTextIter is;

    set_menu_states(FALSE);
    getprefs();
    makecombo();
    assert(NULL != (w = GTK_WIDGET (gtk_builder_get_object (
                                        wbt->builder, "logcombo1"))));
    gtk_combo_box_set_active(GTK_COMBO_BOX(w), 0);
    wbt->serfd = -1;
    assert(NULL != (wbt->tv= GTK_WIDGET (gtk_builder_get_object (
                                             wbt->builder,"gps_text"))));
    wbt->tb = gtk_text_view_get_buffer(GTK_TEXT_VIEW(wbt->tv));
    gtk_text_buffer_get_start_iter (wbt->tb, &is);
    wbt->eob = gtk_text_buffer_create_mark(wbt->tb, "eob", &is, FALSE);
    wbt->serq = NULL;
    make_accel();
}

int main (int argc, char **argv)
{
    gboolean show_vers = FALSE;
    gboolean verbose=FALSE;

    GError *error = NULL;
    GOptionContext *context;
    GOptionEntry entries[] = {
        { "verbose", 'v', 0, G_OPTION_ARG_NONE, &verbose, _("be verbose"), NULL },
        { "version", 'V', 0, G_OPTION_ARG_NONE, &show_vers,
          _("version info"), NULL },
        { NULL }
    };

    bindtextdomain (GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
    bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
    textdomain (GETTEXT_PACKAGE);

    context = g_option_context_new (_("- configure wbt-201 GPS"));
    g_option_context_set_ignore_unknown_options(context,FALSE);
    g_option_context_add_main_entries (context, entries,NULL);
    if(g_option_context_parse (context, &argc, &argv,&error) == FALSE)
    {        if (error != NULL)        {
            fprintf (stderr, "%s\n", error->message);
            g_error_free (error);
            exit(1);
        }
    }

    if(show_vers)
    {
        fputs(PACKAGE " v" VERSION "\n", stderr);
        exit(0);
    }

    wbt = g_malloc0(sizeof(*wbt));
    wbt->verbose = verbose;
    setlocale (LC_ALL, "");
    gtk_init (&argc, &argv);

    wbt->builder = gtk_builder_new ();
    if (!gtk_builder_add_from_file (wbt->builder,
                                    PACKAGE_DATA_DIR "/" "gtk-g-rays2.ui",
                                    &error))
    {
        g_warning ("Couldn't load builder file: %s", error->message);
        g_error_free (error);
        return 0;
    }
   gtk_builder_connect_signals (wbt->builder, NULL);
   setup();
   setup_signals();
   gtk_main ();

   return 0;
}
